### PR TITLE
2316 refactor mocha tests to qunit

### PR DESCRIPTION
**Pull request by candunaj**
Removed mocha. Added Qunit. Refactored unit tests.

### CODE REVIEW

File: packages/ember-simple-auth/tests/index.html [hunk 1]
```diff
@@ -1,33 +1,43 @@
...
-    <script src="{{rootURL}}assets/vendor.js"></script>
-    <script src="{{rootURL}}assets/test-support.js"></script>
-    <script src="{{rootURL}}assets/dummy.js"></script>
-    <script src="{{rootURL}}assets/tests.js"></script>
-
-    {{content-for "body-footer"}}
-    {{content-for "test-body-footer"}}
-  </body>
-</html>
+
```
Comment (BobrImperator) [inline packages/ember-simple-auth/tests/index.html:3]: are changes in this file necessary?

File: packages/ember-simple-auth/package.json [hunk 2]
```diff
@@ -31,11 +31,13 @@
...
   "devDependencies": {
     "@ember/optional-features": "^1.3.0",
+    "@ember/test-helpers": "^2.4.2",
+    "@embroider/test-setup": "^0.43.5",
     "@glimmer/component": "^1.0.0",
     "@glimmer/tracking": "^1.0.0",
     "babel-eslint": "^10.1.0",
     "broccoli-asset-rev": "^3.0.0",
-    "chai": "^4.1.0",
+    "chai": "^4.3.4",
```
Comment (BobrImperator) [inline packages/ember-simple-auth/package.json:40]: chai is a part of mocha, so we want to remove it as well.

**Review by BobrImperator — CHANGES_REQUESTED** [review]

File: packages/ember-simple-auth/package.json [hunk 3]
```diff
@@ -31,11 +31,13 @@
   },
   "devDependencies": {
     "@ember/optional-features": "^1.3.0",
+    "@ember/test-helpers": "^2.4.2",
+    "@embroider/test-setup": "^0.43.5",
```
Comment (BobrImperator) [inline packages/ember-simple-auth/package.json:35]: Embroider test setup is not needed for the migration

File: packages/ember-simple-auth/package.json [hunk 4]
```diff
@@ -31,11 +31,13 @@
...
     "@glimmer/component": "^1.0.0",
     "@glimmer/tracking": "^1.0.0",
     "babel-eslint": "^10.1.0",
     "broccoli-asset-rev": "^3.0.0",
-    "chai": "^4.1.0",
+    "chai": "^4.3.4",
     "ember-cli": "~3.18.0",
     "ember-cli-blueprint-test-helpers": "^0.19.2",
     "ember-cli-chai": "^0.5.0",
```
Comment (BobrImperator) [inline packages/ember-simple-auth/package.json:43]: chai ✂️

File: packages/ember-simple-auth/tests/index.html (hunk 1 above)
Comment (candunaj) [reply packages/ember-simple-auth/tests/index.html:3]: Change in this file fixed route /tests when you run it with "ember s", so I believe that the change in this file is valid.

Comment (BobrImperator) [reply packages/ember-simple-auth/tests/index.html:3]: Now I see, qunit needs some element it can hook into, so we should probably just go back to the old `index.html` file and add this below test-body block:

```html
  <div id="qunit"></div>
  <div id="qunit-fixture">
    <div id="ember-testing-container">
      <div id="ember-testing"></div>
    </div>
  </div>
  ```

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -56,8 +56,6 @@ jobs:
           - classic-test-app
           - test-app
         test-suite:
-          - test:one ember-3.0
-          - test:one ember-lts-3.4
```
Comment (marcoow) [inline .github/workflows/ci.yml:60]: Why does this change? Can we not support 3.0 and 3.4 when using QUnit?

File: packages/classic-test-app/package.json [hunk 6]
```diff
@@ -70,4 +68,4 @@
     "cssstyle": "~1.2.2",
     "jsdom": "^11.10.0"
   }
-}
+}
```
Comment (marcoow) [inline packages/classic-test-app/package.json:71]:
```suggestion
}

```

Comment (marcoow) [reply packages/classic-test-app/package.json:71]: this seems unrelated anyway

File: packages/classic-test-app/tests/.eslintrc.js [hunk 7]
```diff
@@ -2,8 +2,6 @@
 
 module.exports = {
   extends: [
-    'simplabs/configs/ember-mocha',
-    'simplabs/plugins/mocha',
```
Comment (marcoow) [inline packages/classic-test-app/tests/.eslintrc.js:6]: Should this not be changed to `"plugin:qunit/recommended"`?

File: packages/ember-simple-auth/.eslintrc.js [hunk 8]
```diff
@@ -57,6 +57,11 @@ module.exports = {
         node: true,
         mocha: true
       }
-    }
+    },
+    {
+      // test files
+      files: ['tests/**/*-test.{js,ts}'],
+      extends: ['plugin:qunit/recommended'],
+    },
```
Comment (marcoow) [inline packages/ember-simple-auth/.eslintrc.js:65]: See above – I assume this should go into the config that previously contained the Mocha linting settings?

File: packages/ember-simple-auth/addon/services/session.js [hunk 9]
```diff
@@ -1,4 +1,4 @@
-import { alias, oneWay } from '@ember/object/computed';
+import { alias, readOnly } from '@ember/object/computed';
```
Comment (marcoow) [inline packages/ember-simple-auth/addon/services/session.js:1]: This seems unrelated?

File: packages/ember-simple-auth/package.json [hunk 10]
```diff
@@ -94,4 +93,4 @@
   "peerDependencies": {
     "ember-fetch": "^8.0.1"
   }
-}
+}
```
Comment (marcoow) [inline packages/ember-simple-auth/package.json:96]: This seems unrelated – can you revert?

File: packages/ember-simple-auth/tests/.eslintrc.js [hunk 11]
```diff
@@ -2,8 +2,6 @@
 
 module.exports = {
   extends: [
-    'simplabs/configs/ember-mocha',
-    'simplabs/plugins/mocha',
```
Comment (marcoow) [inline packages/ember-simple-auth/tests/.eslintrc.js:6]: see aboev

File: packages/ember-simple-auth/tests/unit/authenticators/oauth2-password-grant-test.js [hunk 12]
```diff
@@ -315,108 +327,123 @@ describe('OAuth2PasswordGrantAuthenticator', () => {
...
+          assert.expect(1);
           server.post('/revoke', (request) => {
             let { requestBody } = request;
             let body = parsePostData(requestBody);
 
-            expect(body).to.eql({
-              'token_type_hint': 'refresh_token',
-              'token': 'refresh token!'
-            });
+            if (body.token_type_hint === 'refresh_token') {
```
Comment (marcoow) [inline packages/ember-simple-auth/tests/unit/authenticators/oauth2-password-grant-test.js:353]: Why is this necessary? It ends up changing the test as well?

File: packages/ember-simple-auth/tests/unit/authenticators/oauth2-password-grant-test.js [hunk 13]
```diff
@@ -315,108 +327,123 @@ describe('OAuth2PasswordGrantAuthenticator', () => {
...
+      assert.expect(1);
       server.post('/token', (request) => {
         let { requestBody } = request;
         let body = parsePostData(requestBody);
 
-        expect(body).to.eql({
-          'grant_type': 'refresh_token',
-          'refresh_token': 'refresh token!'
-        });
+        if (body.grant_type === 'refresh_token') {
```
Comment (marcoow) [inline packages/ember-simple-auth/tests/unit/authenticators/oauth2-password-grant-test.js:397]: see above – why is this necessary?

File: packages/ember-simple-auth/tests/unit/internal-session-store-test.js [hunk 14]
```diff
@@ -1,40 +1,39 @@
...
 
-  describe('session store injection', function() {
-    afterEach(function() {
+  module('session store injection', function(hooks) {
+    hooks.afterEach(function() {
       Ember.testing = true; // eslint-disable-line ember/no-ember-testing-in-module-scope
     });
 
-    it('looks up the test session store when Ember.testing true', function() {
+    test('looks up the test session store when ENV.environment true', function(assert) {
```
Comment (marcoow) [inline packages/ember-simple-auth/tests/unit/internal-session-store-test.js:25]:
```suggestion
    test('looks up the test session store when Ember.testing true', function(assert) {
```

Not sure why this changes?

Comment (marcoow) [reply packages/ember-simple-auth/tests/unit/internal-session-store-test.js:25]: This is only one example case of course – there are plenty of other similar changes as well?

File: packages/ember-simple-auth/tests/unit/internal-session-test.js [hunk 15]
```diff
@@ -23,307 +22,343 @@ describe('InternalSession', () => {
...
       });
 
-      it('is not authenticated', async function() {
-        await session.restore();
+      test('is not authenticated', async function(assert) {
+        try {
+          await session.restore();
+        } catch (e) {
+          // ignore
+        }
```
Comment (marcoow) [inline packages/ember-simple-auth/tests/unit/internal-session-test.js:243]: Can we remove the `try`/`catch`? I assume we want this to raise if `restore` raises an error?

File: packages/ember-simple-auth/tests/unit/internal-session-test.js [hunk 16]
```diff
@@ -23,307 +22,343 @@ describe('InternalSession', () => {
...
```
Comment (marcoow) [inline packages/ember-simple-auth/tests/unit/internal-session-test.js:259]: see above – not sure why we're hiding the error?

File: packages/ember-simple-auth/tests/unit/internal-session-test.js [hunk 17]
```diff
@@ -332,471 +367,533 @@ describe('InternalSession', () => {
...
       });
 
-      it('is not authenticated', async function() {
-        await session.authenticate('authenticator:test');
+      test('is not authenticated', async function(assert) {
+        try {
+          await session.authenticate('authenticator:test');
+        } catch (e) {
+          // ignore
+        }
```
Comment (marcoow) [inline packages/ember-simple-auth/tests/unit/internal-session-test.js:398]: see above

File: packages/ember-simple-auth/tests/unit/internal-session-test.js [hunk 18]
```diff
@@ -332,471 +367,533 @@ describe('InternalSession', () => {
...
+      test('does not process the event', async function(assert) {
         sinon.spy(authenticator, 'restore');
         let restoration = session.restore();
         store.trigger('sessionDataUpdated', { some: 'other property', authenticated: { authenticator: 'authenticator:test' } });
-        await restoration;
+        try {
+          await restoration;
+        } catch (e) {
+          // ignore
+        }
```
Comment (marcoow) [inline packages/ember-simple-auth/tests/unit/internal-session-test.js:619]: see above

File: packages/ember-simple-auth/tests/unit/session-stores/shared/storage-event-handler-behavior.js [hunk 19]
```diff
@@ -1,52 +1,51 @@
+
```
Comment (marcoow) [inline packages/ember-simple-auth/tests/unit/session-stores/shared/storage-event-handler-behavior.js:1]:
```suggestion
```

File: packages/test-app/package.json [hunk 20]
```diff
@@ -73,4 +71,4 @@
     "cssstyle": "~1.2.2",
     "jsdom": "^11.10.0"
   }
-}
+}
```
Comment (marcoow) [inline packages/test-app/package.json:74]:
```suggestion
}

```

File: packages/test-app/config/fastboot-testing.js [hunk 21]
```diff
@@ -1,4 +1,10 @@
-const najax = require('najax');
+let najax;
+try {
+  // eslint-disable-next-line node/no-missing-require
+  najax = require('najax');
+} catch (err) {
+  // ignore the error
```
Comment (marcoow) [inline packages/test-app/config/fastboot-testing.js:6]: Why is this error ignored?

File: packages/test-app/tests/.eslintrc.js [hunk 22]
```diff
@@ -2,8 +2,6 @@
 
 module.exports = {
   extends: [
-    'simplabs/configs/ember-mocha',
-    'simplabs/plugins/mocha',
```
Comment (marcoow) [inline packages/test-app/tests/.eslintrc.js:6]: see above – I think this is where the qunit config should go

File: packages/test-app/tests/index.html [hunk 23]
```diff
@@ -1,33 +1,43 @@
...
+  <script src="{{rootURL}}assets/vendor.js"></script>
+  <script src="{{rootURL}}assets/test-support.js"></script>
+  <script src="{{rootURL}}assets/test-app.js"></script>
+  <script src="{{rootURL}}assets/tests.js"></script>
+
+  {{content-for "body-footer"}}
+  {{content-for "test-body-footer"}}
+</body>
+
+</html>
```
Comment (marcoow) [inline packages/test-app/tests/index.html:43]: Seems like this diff of this file is a bit messed up because of the reformatting (and now missing indentation)?

**Review by marcoow — CHANGES_REQUESTED** [review]
Very nice work :raised_hands:

I found some things that I think still should be addressed but I'm looking forward getting this merged!

One general thing I recognized is that this replaces the nested `describe` blocks with a flat structure with just one `module` and `test`s nested directly into that (at least in some of the test files).Can we use nested `module`s instead? I feel a flat list of tests is much harder to navigate (plus, in a nested structure it's much easier to build up preconditions for tests in nested blocks incrementally).

File: .github/workflows/ci.yml (hunk 5 above)
Comment (candunaj) [reply .github/workflows/ci.yml:60]: Sorry, I wanted to write ember-qunit 5. ember-qunit 5 supports ember 3.8+. 
https://github.com/emberjs/ember-qunit

File: packages/ember-simple-auth/addon/services/session.js (hunk 9 above)
Comment (candunaj) [reply packages/ember-simple-auth/addon/services/session.js:1]: There are 2 unit tests which expect three properties in session service to be readOnly and fail when set. So I have changed properties isAuthenticated, data and store to readOnly.

Comment (marcoow) [reply packages/ember-simple-auth/addon/services/session.js:1]: But `oneWay` should be the same? My point is migrating the tests from Mocha to Qunit shouldn't require any changes in the tested code?

File: .github/workflows/ci.yml (hunk 5 above)
Comment (marcoow) [reply .github/workflows/ci.yml:60]: Can we use a version of ember-qunit that supports Ember 3.0 as well? I'd rather upgrade to ember-qunit 5 (and drop support for Ember < 3.8) in a separate PR then.

Comment (candunaj) [reply .github/workflows/ci.yml:60]: Yes, I can use ember-qunit 4 with some change in the source code. I will do it.

**Comment by BobrImperator** [comment]
I went through the changes and there isn't anything from my side apart from what Marco's pointed out.
Although, I'm not sure if we were supposed to change `test` and `classic-test`apps as well, thought that just ESA package refactor is enough.

File: packages/ember-simple-auth/addon/services/session.js (hunk 9 above)
Comment (candunaj) [reply packages/ember-simple-auth/addon/services/session.js:1]: oneWay will not throw an error when set. I have found the bug in mocha tests so the bug was always there.

```
it('is read-only', function() {
      expect(() => {
        sessionService.set('store', 'some other store');
      }).to.throw;
    });
```
it should be
```
}).to.throw();
```
Change is authenticated from oneWay to readOnly has fixed 3 tests, but it can be considered as breaking change. So maybe we should remove these tests?

Comment (marcoow) [reply packages/ember-simple-auth/addon/services/session.js:1]: Hm, we should definitely move that into a separate PR to not also mix actual changes to the behavior into this – it's already a huge PR (necessarily) so I'd keep it as limited as possible :)

File: packages/classic-test-app/package.json [hunk 24]
```diff
@@ -14,24 +14,22 @@
     "test:fastboot": "ember fastboot:test"
   },
   "devDependencies": {
+    "@ember/jquery": "^2.0.0",
```
Comment (marcoow) [inline packages/classic-test-app/package.json:17]: Is this related?

File: packages/ember-simple-auth/tests/unit/authenticators/oauth2-password-grant-test.js (hunk 12 above)
Comment (marcoow) [reply packages/ember-simple-auth/tests/unit/authenticators/oauth2-password-grant-test.js:353]: @candunaj I assume the `if` could just be dropped?

Comment (marcoow) [reply packages/ember-simple-auth/tests/unit/authenticators/oauth2-password-grant-test.js:353]: @candunaj didn't you extract this into a separate PR?

**Review by marcoow — APPROVED** [review]
great job 🎉

**Review by BobrImperator — APPROVED** [review]
